XMonad/Layout/ByWidth.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE FlexibleInstances, MultiParamTypeClasses #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  XMonad.Layout.ByWidth
+-- Description :  Configure layouts based on the width of your screen.
+-- Copyright   :  (c) Edward Z. Yang
+-- License     :  BSD-style (see LICENSE)
+--
+-- Maintainer  :  <ezyang@cs.stanford.edu>
+-- Stability   :  unstable
+-- Portability :  unportable
+--
+-- Configure layouts based on the width of your screen; use your
+-- favorite multi-column layout for wide screens and a full-screen
+-- layout for small ones.
+-----------------------------------------------------------------------------
+
+module XMonad.Layout.ByWidth
+    ( -- * Usage
+      -- $usage
+      PerScreen,
+      ifWider
+    ) where
+
+import XMonad
+import qualified XMonad.StackSet as W
+
+import XMonad.Prelude (fromMaybe)
+
+-- $usage
+-- You can use this module by importing it into your ~\/.xmonad\/xmonad.hs file:
+--
+-- > import XMonad.Layout.ByWidth
+--
+-- and modifying your layoutHook as follows (for example):
+--
+-- > layoutHook = ifWider 1280 (Tall 1 (3/100) (1/2) ||| Full) Full
+--
+-- Replace any of the layouts with any arbitrarily complicated layout.
+-- 'ifWider' can also be used inside other layout combinators.
+--
+-- For backward compatibility, the type constructor is still called 'PerScreen'.
+
+ifWider :: (LayoutClass l1 a, LayoutClass l2 a)
+               => Dimension   -- ^ target screen width
+               -> l1 a        -- ^ layout to use when the screen is wide enough
+               -> l2 a        -- ^ layout to use otherwise
+               -> PerScreen l1 l2 a
+ifWider w = PerScreen w False
+
+data PerScreen l1 l2 a = PerScreen Dimension Bool (l1 a) (l2 a) deriving (Read, Show)
+
+-- | Construct new PerScreen values with possibly modified layouts.
+mkNewPerScreenT :: PerScreen l1 l2 a -> Maybe (l1 a) ->
+                 PerScreen l1 l2 a
+mkNewPerScreenT (PerScreen w _ lt lf) mlt' =
+    (\lt' -> PerScreen w True lt' lf) $ fromMaybe lt mlt'
+
+mkNewPerScreenF :: PerScreen l1 l2 a -> Maybe (l2 a) ->
+                 PerScreen l1 l2 a
+mkNewPerScreenF (PerScreen w _ lt lf) mlf' =
+    PerScreen w False lt $ fromMaybe lf mlf'
+
+instance (LayoutClass l1 a, LayoutClass l2 a, Show a) => LayoutClass (PerScreen l1 l2) a where
+    runLayout (W.Workspace i p@(PerScreen w _ lt lf) ms) r
+        | rect_width r > w    = do (wrs, mlt') <- runLayout (W.Workspace i lt ms) r
+                                   return (wrs, Just $ mkNewPerScreenT p mlt')
+        | otherwise           = do (wrs, mlt') <- runLayout (W.Workspace i lf ms) r
```

Comment:
Just doing `runLayout` with the other layout whenever the condition changes will leave decorations (and other stuff like this) around. In other words, this suffers from https://github.com/xmonad/xmonad-contrib/issues/75. You need to send `Hide` to the last layout first before `runLayout`ing the new one, like this: https://github.com/xmonad/xmonad-contrib/pull/582#issuecomment-898495845

Also, you need to handle `ReleaseResources` and pass it on to both layouts.

Reply:
Oh, now I see you're just moving stuff around and this is still the same code that was there. So I guess that's an argument for not fixing it in the scope of this PR.

That then leads me to a thought whether it wouldn't be better to not hijack the `PerScreen` name and just add an `X.L.OnScreen` module… Deprecating `PerScreen` in favor of `ByWidth` or `IfWidth` can stay, of course.

Oh, and related to my previous comment about #75 - the plan was to implement #582 generic enough so that `IfWidth` can be implemented using it and thus get #75 fixed for free. So we may want to limit the exports a bit… although we can always use `pattern` for backwards compat.

Reply:
The main reason I reused `PerScreen` is consistency with `PerWorkspace` and friends.

And yes, the only changes I made to `ByWidth` were renames; everything else is the same, including the weirdnesses @TheMC47 noticed. I was actually somewhat afraid to touch them, because @ezyang is more of a Haskeller than I am and I tend to assume "fancy" code has some reason not visible to me.

Reply:
Re `Hide` and `ReleaseResources`: I did think about that, but thought we had a more global plan to deal with those somewhere. I can add that. What worries _me_ is I don't think it scales; are there, or will there at some point be, more "global" messages that need to be dealt with by individual layouts?